### ISSUE TITLE
Large Discrepancy Between AIMET QuantSim and QNN Runtime Outputs Despite High QuantSim Accuracy

### ISSUE THREAD

**Issue by excllent123** [quic/aimet#4060]

**Comment by excllent123** [comment]
We would like to share an observation regarding the impact of different upsampling strategies when deploying our custom UNetRes model (details below) via QNN. Specifically, we compared two upsampling implementations:

1. **`upsample_upconv`**: Nearest-neighbor upsampling followed by a 3×3 convolution  
2. **`pixel_shuffle`**: Depth-to-space based upsampling (also known as sub-pixel convolution)

### Key Findings:
When using **`upsample_upconv`**, the quantized model (`qnn-context-binary`) achieves excellent fidelity relative to the floating-point reference:

- **MAE**: 7.69 DN  
- **PSNR**: 76.24 dB  
- **>5 DN error ratio**: only **0.578%**

However, switching to **`pixel_shuffle`** leads to a significant degradation in reconstruction quality:

- **MAE**: 17.76 DN (**+130% increase**)  
- **PSNR**: 69.52 dB (**−6.7 dB drop**)  
- **>5 DN error ratio**: **14.07%** (**24× higher**)

Region-wise mean comparisons also show consistently larger deviations across all quadrants with `pixel_shuffle`.

### Model Architecture Summary:
```python
class UNetRes(nn.Module):
    def __init__(self, in_nc=1, out_nc=1, nc=[16, 16, 32, 32], nb=[2, 3, 3, 3], act_mode='r', downsample_mode='strideconv', upsample_mode='upconv'):
        super().__init__()
        self.m_head = Conv3X3(in_nc, nc[0],act_type='relu')
        downsample_block = B.downsample_strideconv

        self.m_down1 = B.sequential(*[Conv3X3(nc[0], nc[0], act_type='relu') for _ in range(nb[0])], downsample_block(nc[0], nc[1], bias=False, mode='2'))
        self.m_down2 = B.sequential(*[B.ResBlock(nc[1], nc[1], bias=False, mode='C'+act_mode+'C') for _ in range(nb[1])], downsample_block(nc[1], nc[2], bias=False, mode='2'))
        self.m_down3 = B.sequential(*[B.ResBlock(nc[2], nc[2], bias=False, mode='C'+act_mode+'C') for _ in range(nb[2])], downsample_block(nc[2], nc[3], bias=False, mode='2'))

        self.m_body = B.sequential(*[B.ResBlock(nc[3], nc[3], bias=False, mode='C'+act_mode+'C') for _ in range(nb[3])])

        upsample_block = B.upsample_upconv

        self.m_up3 = B.sequential(upsample_block(nc[3], nc[2], bias=False, mode='2'), *[B.ResBlock(nc[2], nc[2], bias=False, mode='C'+act_mode+'C') for _ in range(nb[2])])
        self.m_up2 = B.sequential(upsample_block(nc[2], nc[1], bias=False, mode='2'), *[B.ResBlock(nc[1], nc[1], bias=False, mode='C'+act_mode+'C') for _ in range(nb[1])])
        self.m_up1 = B.sequential(upsample_block(nc[1], nc[0], bias=False, mode='2'), *[Conv3X3(nc[0], nc[0], act_type='relu') for _ in range(nb[0])])

        self.m_tail = nn.Sequential(
            nn.Conv2d(16, 1, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False),
            nn.Hardtanh(0.0, 1.0)  
        )
        
    def forward(self, x0):
        x0 = self.m_head(x0)
        x1 = self.m_down1(x0)
        x2 = self.m_down2(x1)
        x3 = self.m_down3(x2)
        x = self.m_body(x3)
        x = self.m_up3(x+x3)
        x = self.m_up2(x+x2)
        x = self.m_up1(x+x1)
        x = self.m_tail(x)
        return x
```
The network uses residual blocks in the encoder/decoder paths and ReLU activations in the shallow layers.

### Request:
Given these results, we plan to retain `upsample_upconv` for deployment. However, we’d appreciate your insight on:
- Whether this behavior is expected due to how QNN handles `pixel_shuffle` (e.g., precision loss in index manipulation or interpolation)?
- Any known best practices or constraints around `pixel_shuffle` in QNN quantization pipelines?

**Comment by bhushan23** [comment]
@excllent123 indeed excellent questions and thank you for being thorough on sharing repro, examples and configuration

> Why does the QNN-compiled binary exhibit large internal discrepancies—even though AIMET QuantSim shows near-perfect fidelity to FP32?
Is this behavior expected given the model architecture (e.g., deep cumulative additions, PixelShuffle, Hardtanh)?
Main reason behind this is how this is deep cumulative addition on certain architectures.
Sim to target is not expected to match bit-exact and depending on model and activations, this can add up in certain cases.

>How can we effectively debug and pinpoint the source of this sim-to-target mismatch?
Are there recommended tools or strategies (e.g., layer-by-layer dump, scale alignment checks, or HTP-specific limitations) to isolate whether the issue stems from:
>>Numerical precision loss in accumulators?

Using tools to for layer-by-layer dump is recommended to see where does discrepancy begin
I also prefer binary search over onnx to validate accumulated error e.g. you can add multiple intermediate layers as model output to capture in one shot.

>>Differences in operator implementation (e.g., PixelShuffle, Hardtanh)?

If operator implementation should not be different from standard runtime e.g. onnxruntime-cpu and if we can proof this with small toy model, we get this to QAIRT teams attention to fix at highest priority

>>Scale propagation errors across skip connections? 

Ideally, this should not happen as this will be captured during simulation time.

**Comment by quic-kyunggeu** [comment]
Hi @excllent123, thanks for excellent repro 👍

The low cosine similarity between intermediate tensors looks like a false alarm due to a bug in QAIRT accuracy debugger. It's trying to compare NCHW tensors (onnx) to NHWC tenosrs (dlc) without considering the tensor layout difference.
As you may already know, ONNX Conv uses NCHW layout while DLC uses NHWC. However, QAIRT accuracy debugger seems to simply compare the flattened vectors of the two tensors without aligning the tensor layout. This hypothesis explains two things.
1. Final output `output`, which has no channel axis, shows ~1.0 cosine similarity
2. Tensor `_m_tail_m_tail_1_DepthToSpace_output_0`, which is the one previous to the final output, also shows ~1.0 cosine similarity because C=1; all other tensors with C != 1 show near-zero cosine similarity.

To summarize, this is a false alarm due to QAIRT accuracy debugger. We will file an internal ticket to QAIRT team to fix this issue.

**Comment by quic-kyunggeu** [comment]
@excllent123  Follow-up

According to QAIRT debugger dev team, you need to additionally pass a --dlc_file to --verification command to make the debugger aware of tensor layout difference.
```bash
qairt-accuracy-debugger --verification \
  --comparators cosine mse \
  --inference_tensor working_directory/framework_runner/2026-01-09_09-16-23 \
  --reference_tensor working_directory/framework_runner/2026-01-09_09-16-23 \
  **--dlc_file Unet_Res.dlc**
```
